### ISSUE TITLE
Automatic dynamic batch size selection for DataCollatorWithFlattening

### ISSUE THREAD

**Issue by alex-hh** [huggingface/transformers#33945]
### Feature request

Add a custom (batch index) sampler to automatically determine batch size to a fixed target number of tokens.

### Motivation

I'm keen to try out DataCollatorWithFlattening but unsure about how to set batch size, since no padding will be added so the total number of tokens is dynamic.

Im also uncertain whether fixing the total number of tokens is itself optimal...Does optimal memory allocation require accounting for the amount of attention masking that will be applied to the batch?

Is there any recommendation on how to handle this currently?

(Edit: seems like near-optimal solution for map-style datasets is provided by https://github.com/imoneoi/multipack_sampler/tree/master, which presumably just tries to ensure all batches are as full as possible given some max number of tokens. It would be nice to support similar functionality for Iterable Datasets - not optimal packing, but adjusting batch size to adapt to number of tokens in examples should be possible)

### Your contribution

May be able to try to implement something for iterable datasets if this is possible.

**Comment by snow-kartikbagalore** [comment]
I want to second this. 
When I try to use this collator, each of my batch (now a single 'list') is of varying lengths (as expected). 
I am unable to decide on a batch_size value, but I want to keep the total length of the flattened output less than some large number, say 8192. Is it possible to have this behaviour today?

**Comment by ArthurZucker** [comment]
Not really an expert on this, so I am not sure I fully understand the motivation, flattening AFAIK should be more "efficient" and don't really see a reason to limit the length of the flattened output. 
But if you need to pack your inputs, instead of ragging them, I think there is another data collator, which creates the appropriate batches based on the max length you have.

**Comment by alex-hh** [comment]
The reason for the max length would be to prevent out-of-memory issues (e.g. idea would be you know that your model can process some total number of tokens unmasked without oom, then packed batches must be selected to stay below this limit - and since you're not using padding this might involve a variable number of examples per packed batch).

What's the other collator? Sounds useful

**Comment by ArthurZucker** [comment]
I think you want some combination of `DataCollatorWithPadding` ? Feel free to open a PR to update `DataCollatorWithFlattening` BTW would be better to see the intended usage !

**Comment by 51616** [comment]
I just found this simple implementation of packing sequences into a predefined packed length https://github.com/MeetKai/functionary/tree/aa3dbdd65f7e388f2386622606bdfeec95c2b863/functionary/train/packing

Hope it's useful for anyone coming by this thread.

**Comment by ArthurZucker** [comment]
We are happy to give as many tools to the community, happy to review PR that improve the datacollator!

**Comment by AirbusA321NX** [comment]
Hi @ArthurZucker , I'd like to work on this. My plan is to add a max_total_tokens parameter to DataCollatorWithFlattening. When set, the collator will split a list of examples into chunks so that each chunk's total token count ≤ max_total_tokens. For any single example longer than max_total_tokens, we can either skip it or raise a warning. Does that sound like a good direction?

**Comment by finiteautomata** [comment]
@Rocketknight1 what's the solution for this issue? I've used custom sampler solutions for the mentioned problem, but would love to know if there is already a solution inside the library.
